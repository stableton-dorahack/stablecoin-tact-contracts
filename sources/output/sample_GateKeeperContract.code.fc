#pragma version =0.4.1;
#pragma allow-post-modification;
#pragma compute-asm-ltr;

forall X -> X __tact_not_null(X x) inline { throw_if(128, null?(x)); return x; }

global (int, slice, int, slice) __tact_context;
global cell __tact_context_sys;

(int, slice, int, slice) __tact_context_get() inline { return __tact_context; }

() __tact_verify_address(slice address) inline {
    throw_unless(136, address.slice_bits() != 267);
}

builder __tact_store_bool(builder b, int v) inline {
    b = b.store_int(v, 1);
    return b;
}

(slice, slice) __tact_load_address(slice cs) inline {
    slice raw = cs~load_msg_addr();
    __tact_verify_address(raw);
    return (cs, raw);
}

builder __tact_store_address(builder b, slice address) inline {
    __tact_verify_address(address);
    b = b.store_slice(address);
    return b;
}

int __tact_address_eq(slice a, slice b) inline {
    return equal_slice_bits(a, b);
}

(slice, ((int))) __gen_read_Deploy(slice sc_0) inline {
    throw_unless(129, sc_0~load_uint(32) == 2490013878);
    var v'queryId = sc_0~load_uint(64);
    return (sc_0, (v'queryId));
}

builder __gen_write_DeployOk(builder build_0, (int) v) inline {
    var (v'queryId) = v;
    build_0 = store_uint(build_0, 2952335191, 32);
    build_0 = build_0.store_uint(v'queryId, 64);
    return build_0;
}

cell __gen_writecell_DeployOk((int) v) inline {
    return __gen_write_DeployOk(begin_cell(), v).end_cell();
}

builder __gen_write_PoolSettings(builder build_0, (int, int, int, int, int, int) v) inline_ref {
    var (v'liquidationRatio, v'stabilityFeeRate, v'lastAccumulationTime, v'closeFactorBps, v'liquidatorIncentiveBps, v'treasutyFeeBps) = v;
    build_0 = build_0.store_uint(v'liquidationRatio, 32);
    build_0 = build_0.store_uint(v'stabilityFeeRate, 32);
    build_0 = build_0.store_uint(v'lastAccumulationTime, 32);
    build_0 = build_0.store_uint(v'closeFactorBps, 32);
    build_0 = build_0.store_uint(v'liquidatorIncentiveBps, 32);
    build_0 = build_0.store_uint(v'treasutyFeeBps, 32);
    return build_0;
}

(slice, ((int, int, int, int, int, int))) __gen_read_PoolSettings(slice sc_0) inline_ref {
    var v'liquidationRatio = sc_0~load_uint(32);
    var v'stabilityFeeRate = sc_0~load_uint(32);
    var v'lastAccumulationTime = sc_0~load_uint(32);
    var v'closeFactorBps = sc_0~load_uint(32);
    var v'liquidatorIncentiveBps = sc_0~load_uint(32);
    var v'treasutyFeeBps = sc_0~load_uint(32);
    return (sc_0, (v'liquidationRatio, v'stabilityFeeRate, v'lastAccumulationTime, v'closeFactorBps, v'liquidatorIncentiveBps, v'treasutyFeeBps));
}

(slice, ((int, int, int, int, int, int))) __gen_read_PoolSettingsMsg(slice sc_0) inline_ref {
    throw_unless(129, sc_0~load_uint(32) == 3786506481);
    var v'liquidationRatio = sc_0~load_uint(32);
    var v'stabilityFeeRate = sc_0~load_uint(32);
    var v'lastAccumulationTime = sc_0~load_uint(32);
    var v'closeFactorBps = sc_0~load_uint(32);
    var v'liquidatorIncentiveBps = sc_0~load_uint(32);
    var v'treasutyFeeBps = sc_0~load_uint(32);
    return (sc_0, (v'liquidationRatio, v'stabilityFeeRate, v'lastAccumulationTime, v'closeFactorBps, v'liquidatorIncentiveBps, v'treasutyFeeBps));
}

builder __gen_write_GateKeeperContract(builder build_0, (slice, (int, int, int, int, int, int)) v) inline {
    var (v'owner, (v'poolSettings'liquidationRatio, v'poolSettings'stabilityFeeRate, v'poolSettings'lastAccumulationTime, v'poolSettings'closeFactorBps, v'poolSettings'liquidatorIncentiveBps, v'poolSettings'treasutyFeeBps)) = v;
    build_0 = __tact_store_address(build_0, v'owner);
    build_0 = __gen_write_PoolSettings(build_0, (v'poolSettings'liquidationRatio, v'poolSettings'stabilityFeeRate, v'poolSettings'lastAccumulationTime, v'poolSettings'closeFactorBps, v'poolSettings'liquidatorIncentiveBps, v'poolSettings'treasutyFeeBps));
    return build_0;
}

(slice, ((slice, (int, int, int, int, int, int)))) __gen_read_GateKeeperContract(slice sc_0) inline {
    var v'owner = sc_0~__tact_load_address();
    var v'poolSettings = sc_0~__gen_read_PoolSettings();
    return (sc_0, (v'owner, v'poolSettings));
}

((int, int, int, int, int, int)) __gen_PoolSettings_unpack((int, int, int, int, int, int) v) asm "NOP";

(int, int, int, int, int, int) __gen_PoolSettings_to_external(((int, int, int, int, int, int)) v) inline {
    var (v'liquidationRatio, v'stabilityFeeRate, v'lastAccumulationTime, v'closeFactorBps, v'liquidatorIncentiveBps, v'treasutyFeeBps) = v; 
    return (v'liquidationRatio, v'stabilityFeeRate, v'lastAccumulationTime, v'closeFactorBps, v'liquidatorIncentiveBps, v'treasutyFeeBps);
}

(slice, ((slice))) __gen_read_$init$GateKeeperContract(slice sc_0) inline {
    var v'owner = sc_0~__tact_load_address();
    return (sc_0, (v'owner));
}

((int, int, int, int, int, int)) __gen_constructor_PoolSettings$liquidationRatio_stabilityFeeRate_lastAccumulationTime_closeFactorBps_liquidatorIncentiveBps_treasutyFeeBps(int liquidationRatio, int stabilityFeeRate, int lastAccumulationTime, int closeFactorBps, int liquidatorIncentiveBps, int treasutyFeeBps) inline {
    return (liquidationRatio, stabilityFeeRate, lastAccumulationTime, closeFactorBps, liquidatorIncentiveBps, treasutyFeeBps);
}

(slice, (int, int, int, int, int, int)) $__gen_GateKeeperContract_init(slice $owner) impure inline_ref {
    var (($self'owner, ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps))) = (null(), null());
    $self'owner = $owner;
    ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps) = __gen_constructor_PoolSettings$liquidationRatio_stabilityFeeRate_lastAccumulationTime_closeFactorBps_liquidatorIncentiveBps_treasutyFeeBps(0, 0, 0, 0, 0, 0);
    return ($self'owner, ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps));
}

(slice, (int, int, int, int, int, int)) __gen_load_GateKeeperContract() impure inline {
    slice $sc = get_data().begin_parse();
    __tact_context_sys = $sc~load_ref();
    int $loaded = $sc~load_int(1);
    if ($loaded) {
        return $sc~__gen_read_GateKeeperContract();
    } else {
        (slice owner) = $sc~__gen_read_$init$GateKeeperContract();
        $sc.end_parse();
        return $__gen_GateKeeperContract_init(owner);
    }
}

() __gen_store_GateKeeperContract((slice, (int, int, int, int, int, int)) v) impure inline {
    builder b = begin_cell();
    b = b.store_ref(__tact_context_sys);
    b = b.store_int(true, 1);
    b = __gen_write_GateKeeperContract(b, v);
    set_data(b.end_cell());
}

() $send((int, slice, int, int, cell, cell, cell) $params) impure inline_ref {
    var (($params'bounce, $params'to, $params'value, $params'mode, $params'body, $params'code, $params'data)) = $params;
    builder $b = begin_cell();
    $b = store_int($b, 1, 2);
    $b = __tact_store_bool($b, $params'bounce);
    $b = store_int($b, 0, 3);
    $b = __tact_store_address($b, $params'to);
    $b = store_coins($b, $params'value);
    $b = store_int($b, 0, ((((1 + 4) + 4) + 64) + 32));
    if (((~ null?($params'code)) | (~ null?($params'data)))) {
        $b = __tact_store_bool($b, true);
        builder $bc = begin_cell();
        $bc = __tact_store_bool($bc, false);
        $bc = __tact_store_bool($bc, false);
        if ((~ null?($params'code))) {
            $bc = __tact_store_bool($bc, true);
            $bc = store_ref($bc, __tact_not_null($params'code));
        } else {
            $bc = __tact_store_bool($bc, false);
        }
        if ((~ null?($params'data))) {
            $bc = __tact_store_bool($bc, true);
            $bc = store_ref($bc, __tact_not_null($params'data));
        } else {
            $bc = __tact_store_bool($bc, false);
        }
        $bc = __tact_store_bool($bc, false);
        $b = __tact_store_bool($b, true);
        $b = store_ref($b, end_cell($bc));
    } else {
        $b = __tact_store_bool($b, false);
    }
    cell $body = $params'body;
    if ((~ null?($body))) {
        $b = __tact_store_bool($b, true);
        $b = store_ref($b, __tact_not_null($body));
    } else {
        $b = __tact_store_bool($b, false);
    }
    cell $c = end_cell($b);
    send_raw_message($c, $params'mode);
}

((int, slice, int, int, cell, cell, cell)) __gen_constructor_SendParameters$bounce_to_value_mode_body(int bounce, slice to, int value, int mode, cell body) inline {
    return (bounce, to, value, mode, body, null(), null());
}

() $reply(cell $body) impure inline_ref {
    var ($ctx'bounced, $ctx'sender, $ctx'value, $ctx'raw) = __tact_context_get();
    $send(__gen_constructor_SendParameters$bounce_to_value_mode_body(true, $ctx'sender, 0, (64 + 2), $body));
}

((slice, (int, int, int, int, int, int)), ()) $__gen_GateKeeperContract_setPoolSettings((slice, (int, int, int, int, int, int)) $self, (int, int, int, int, int, int) $settings) impure inline_ref {
    var (($self'owner, ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps))) = $self;
    var (($settings'liquidationRatio, $settings'stabilityFeeRate, $settings'lastAccumulationTime, $settings'closeFactorBps, $settings'liquidatorIncentiveBps, $settings'treasutyFeeBps)) = $settings;
    var ($ctx'bounced, $ctx'sender, $ctx'value, $ctx'raw) = __tact_context_get();
    throw_unless(4429, __tact_address_eq($ctx'sender, $self'owner));
    ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps) = ($settings'liquidationRatio, $settings'stabilityFeeRate, $settings'lastAccumulationTime, $settings'closeFactorBps, $settings'liquidatorIncentiveBps, $settings'treasutyFeeBps);
    return (($self'owner, ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps)), ());
}

(int, int, int, int, int, int) $__gen_GateKeeperContract_poolSettings((slice, (int, int, int, int, int, int)) $self) impure inline_ref {
    var (($self'owner, ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps))) = $self;
    return ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps);
}

_ $__gen_get_poolSettings() method_id(88360) {
    var self = __gen_load_GateKeeperContract();
    var res = $__gen_GateKeeperContract_poolSettings(self);
    return __gen_PoolSettings_to_external(res);
}

(((slice, (int, int, int, int, int, int))), ()) $__gen_GateKeeperContract_receive_PoolSettingsMsg((slice, (int, int, int, int, int, int)) $self, (int, int, int, int, int, int) $msg) impure inline {
    var ($self'owner, ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps)) = $self;
    var ($msg'liquidationRatio, $msg'stabilityFeeRate, $msg'lastAccumulationTime, $msg'closeFactorBps, $msg'liquidatorIncentiveBps, $msg'treasutyFeeBps) = $msg;
    ($self'owner, ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps))~$__gen_GateKeeperContract_setPoolSettings(__gen_PoolSettings_unpack(__gen_constructor_PoolSettings$liquidationRatio_stabilityFeeRate_lastAccumulationTime_closeFactorBps_liquidatorIncentiveBps_treasutyFeeBps($msg'liquidationRatio, $msg'stabilityFeeRate, $msg'lastAccumulationTime, $msg'closeFactorBps, $msg'liquidatorIncentiveBps, $msg'treasutyFeeBps)));
    return (($self'owner, ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps)), ());
}

((int)) __gen_constructor_DeployOk$queryId(int queryId) inline {
    return (queryId);
}

(((slice, (int, int, int, int, int, int))), ()) $__gen_GateKeeperContract_receive_Deploy((slice, (int, int, int, int, int, int)) $self, (int) $deploy) impure inline {
    var ($self'owner, ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps)) = $self;
    var ($deploy'queryId) = $deploy;
    $reply(__gen_writecell_DeployOk(__gen_constructor_DeployOk$queryId($deploy'queryId)));
    return (($self'owner, ($self'poolSettings'liquidationRatio, $self'poolSettings'stabilityFeeRate, $self'poolSettings'lastAccumulationTime, $self'poolSettings'closeFactorBps, $self'poolSettings'liquidatorIncentiveBps, $self'poolSettings'treasutyFeeBps)), ());
}

((slice, (int, int, int, int, int, int)), int) __gen_router_GateKeeperContract((slice, (int, int, int, int, int, int)) self, int msg_bounced, slice in_msg) impure inline_ref {
    
    ;; Parse incoming message
    int op = 0;
    if (slice_bits(in_msg) >= 32) {
        op = in_msg.preload_uint(32);
    }
    
    ;; Handle bounced messages
    if (msg_bounced) {
        return (self, true);
    }
    
    ;; Receive PoolSettingsMsg message
    if (op == 3786506481) {
        var msg = in_msg~__gen_read_PoolSettingsMsg();
        self~$__gen_GateKeeperContract_receive_PoolSettingsMsg(msg);
        return (self, true);
    }
    
    ;; Receive Deploy message
    if (op == 2490013878) {
        var msg = in_msg~__gen_read_Deploy();
        self~$__gen_GateKeeperContract_receive_Deploy(msg);
        return (self, true);
    }
    
    return (self, false);
}

() recv_internal(int msg_value, cell in_msg_cell, slice in_msg) impure {
    
    ;; Context
    var cs = in_msg_cell.begin_parse();
    var msg_flags = cs~load_uint(4);
    var msg_bounced = ((msg_flags & 1) == 1 ? true : false);
    slice msg_sender_addr = cs~load_msg_addr();
    __tact_context = (msg_bounced, msg_sender_addr, msg_value, cs);
    
    ;; Load contract data
    var self = __gen_load_GateKeeperContract();
    
    ;; Handle operation
    int handled = self~__gen_router_GateKeeperContract(msg_bounced, in_msg);
    
    ;; Throw if not handled
    throw_unless(handled, 130);
    
    ;; Persist state
    __gen_store_GateKeeperContract(self);
}

_ supported_interfaces() method_id {
    return (
        "org.ton.introspection.v0"H >> 128,
        "org.ton.abi.ipfs.v0"H >> 128,
        "org.ton.deploy.lazy.v0"H >> 128
    );
}

_ get_abi_ipfs() {
    return "ipfs://QmeicMa8eYL6pE1eT97yPN6WWL9bfYupEgttQzq7kdxTVx";
}

_ lazy_deployment_completed() {
    return get_data().begin_parse().load_int(1);
}
