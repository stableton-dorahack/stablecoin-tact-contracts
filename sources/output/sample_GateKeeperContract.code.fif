PROGRAM{
  DECLPROC __tact_context_get
  DECLPROC __tact_verify_address
  DECLPROC __tact_load_address
  DECLPROC __tact_store_address
  DECLPROC __tact_address_eq
  DECLPROC __gen_read_ChangeOwner
  DECLPROC __gen_write_PoolSettings
  DECLPROC __gen_read_PoolSettings
  DECLPROC __gen_write_GateKeeperContract
  DECLPROC __gen_read_GateKeeperContract
  DECLPROC __gen_Context_get_sender
  DECLPROC __gen_PoolSettings_to_external
  DECLPROC __gen_read_$init$GateKeeperContract
  DECLPROC $__gen_GateKeeperContract_init
  DECLPROC __gen_load_GateKeeperContract
  DECLPROC __gen_store_GateKeeperContract
  DECLPROC $__gen_GateKeeperContract_poolSettings
  88360 DECLMETHOD $__gen_get_poolSettings
  DECLPROC $__gen_GateKeeperContract_requireOwner
  DECLPROC $__gen_GateKeeperContract_owner
  83229 DECLMETHOD $__gen_get_owner
  DECLPROC $__gen_GateKeeperContract_receive_ChangeOwner
  DECLPROC __gen_router_GateKeeperContract
  DECLPROC recv_internal
  113617 DECLMETHOD supported_interfaces
  DECLPROC get_abi_ipfs
  DECLPROC lazy_deployment_completed
  DECLGLOBVAR __tact_context
  DECLGLOBVAR __tact_context_sys
  __tact_context_get PROCINLINE:<{
    __tact_context GETGLOB
    4 UNTUPLE
  }>
  __tact_verify_address PROCINLINE:<{
    SBITS
    267 PUSHINT
    NEQ
    136 THROWIFNOT
  }>
  __tact_load_address PROCINLINE:<{
    LDMSGADDR
    SWAP
  }>
  __tact_store_address PROCINLINE:<{
    STSLICER
  }>
  __tact_address_eq PROCINLINE:<{
    SDEQ
  }>
  __gen_read_ChangeOwner PROCINLINE:<{
    32 LDU
    SWAP
    256331011 PUSHINT
    EQUAL
    129 THROWIFNOT
    __tact_load_address INLINECALLDICT
  }>
  __gen_write_PoolSettings PROCREF:<{
    s5 s6 XCHG2
    32 STU
    s1 s3 XCHG
    32 STU
    32 STU
    32 STU
    32 STU
    32 STU
  }>
  __gen_read_PoolSettings PROCREF:<{
    32 LDU
    32 LDU
    32 LDU
    32 LDU
    32 LDU
    32 LDU
    6 -ROLL
  }>
  __gen_write_GateKeeperContract PROCINLINE:<{
    s7 s6 XCHG2
    __tact_store_address INLINECALLDICT
    s0 s6 XCHG
    5 ROLL
    __gen_write_PoolSettings INLINECALLDICT
  }>
  __gen_read_GateKeeperContract PROCINLINE:<{
    __tact_load_address INLINECALLDICT
    SWAP
    __gen_read_PoolSettings INLINECALLDICT
    s6 s7 XCHG
  }>
  __gen_Context_get_sender PROCINLINE:<{
    s2 s3 XCHG
    3 BLKDROP
  }>
  __gen_PoolSettings_to_external PROCINLINE:<{
  }>
  __gen_read_$init$GateKeeperContract PROCINLINE:<{
    __tact_load_address INLINECALLDICT
  }>
  $__gen_GateKeeperContract_init PROCREF:<{
    PUSHNULL
  }>
  __gen_load_GateKeeperContract PROCINLINE:<{
    c4 PUSH
    CTOS
    LDREF
    SWAP
    __tact_context_sys SETGLOB
    1 LDI
    SWAP
    IF:<{
      __gen_read_GateKeeperContract INLINECALLDICT
      1 7 BLKDROP2
    }>ELSE<{
      __gen_read_$init$GateKeeperContract INLINECALLDICT
      SWAP
      ENDS
      $__gen_GateKeeperContract_init INLINECALLDICT
    }>
  }>
  __gen_store_GateKeeperContract PROCINLINE:<{
    NEWC
    __tact_context_sys GETGLOB
    SWAP
    STREF
    TRUE
    SWAP
    1 STI
    7 -ROLL
    __gen_write_GateKeeperContract INLINECALLDICT
    ENDC
    c4 POP
  }>
  $__gen_GateKeeperContract_poolSettings PROCREF:<{
    1 6 BLKDROP2
  }>
  $__gen_get_poolSettings PROC:<{
    __gen_load_GateKeeperContract INLINECALLDICT
    $__gen_GateKeeperContract_poolSettings INLINECALLDICT
    __gen_PoolSettings_to_external INLINECALLDICT
  }>
  $__gen_GateKeeperContract_requireOwner PROCREF:<{
    __tact_context_get INLINECALLDICT
    __gen_Context_get_sender INLINECALLDICT
    s7 PUSH
    __tact_address_eq INLINECALLDICT
    132 THROWIFNOT
  }>
  $__gen_GateKeeperContract_owner PROCREF:<{
    6 BLKDROP
  }>
  $__gen_get_owner PROC:<{
    __gen_load_GateKeeperContract INLINECALLDICT
    $__gen_GateKeeperContract_owner INLINECALLDICT
  }>
  $__gen_GateKeeperContract_receive_ChangeOwner PROCINLINE:<{
    7 -ROLL
    $__gen_GateKeeperContract_requireOwner INLINECALLDICT
    1 6 BLKDROP2
  }>
  __gen_router_GateKeeperContract PROCREF:<{
    0 PUSHINT
    OVER
    SBITS
    31 GTINT
    IF:<{
      DROP
      DUP
      32 PLDU
    }>
    s0 s2 XCHG
    IFJMP:<{
      2DROP
      TRUE
    }>
    SWAP
    256331011 PUSHINT
    EQUAL
    IFJMP:<{
      __gen_read_ChangeOwner INLINECALLDICT
      NIP
      $__gen_GateKeeperContract_receive_ChangeOwner INLINECALLDICT
      TRUE
    }>
    DROP
    FALSE
  }>
  recv_internal PROC:<{
    SWAP
    CTOS
    4 LDU
    SWAP
    1 PUSHINT
    AND
    1 EQINT
    IF:<{
      TRUE
    }>ELSE<{
      FALSE
    }>
    SWAP
    LDMSGADDR
    s2 PUSH
    s5 s5 XCHG2
    4 TUPLE
    __tact_context SETGLOB
    __gen_load_GateKeeperContract INLINECALLDICT
    2 7 BLKSWAP
    __gen_router_GateKeeperContract INLINECALLDICT
    DROP
    __gen_store_GateKeeperContract INLINECALLDICT
  }>
  supported_interfaces PROC:<{
    123515602279859691144772641439386770278 PUSHINT
    209801025412363888721030803524359905849 PUSHINT
    42980537499636128163026532310500881091 PUSHINT
    212791963610352222038671851416501365766 PUSHINT
    86142586315491086060343270784266291122 PUSHINT
  }>
  get_abi_ipfs PROC:<{
    x{697066733a2f2f516d576d4b39376b4e507641626b39504d53566f4e763472654e6e7771444b61534334484535517754706a6b4a70} PUSHSLICE
  }>
  lazy_deployment_completed PROC:<{
    c4 PUSH
    CTOS
    1 LDI
    SWAP
  }>
}END>c
