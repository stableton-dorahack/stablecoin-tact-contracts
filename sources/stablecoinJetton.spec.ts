import { StablecoinJettonContract } from "./output/stableton_StablecoinJettonContract";
import { Address, beginCell, OpenedContract, Sender, toNano } from "ton";
import { ContractSystem, Logger, Tracker, Treasure } from "ton-emulator";

describe("StablecoinJettonContract", () => {
    let system: ContractSystem;
    let owner: Treasure;
    let user: Treasure;
    let userPosition: Treasure;
    let positionsManager: Treasure;
    let gateKeeper: Treasure;
    let jetton: Treasure;
    let stablecoinMaster: Treasure;
    let stablecoinJettonContract: OpenedContract<StablecoinJettonContract>;
    let track: Tracker;
    let logger: Logger;

    beforeAll(async () => {
        system = await ContractSystem.create();
        owner = system.treasure("owner");
        user = system.treasure("user");
        gateKeeper = system.treasure("gateKeeper");
        positionsManager = system.treasure("positionsManager");
        userPosition = system.treasure("userPosition");
        jetton = system.treasure("jetton");
        stablecoinMaster = system.treasure("stablecoinMaster");

        stablecoinJettonContract = system.open(
            await StablecoinJettonContract.fromInit(stablecoinMaster.address, user.address)
        );

        track = system.track(stablecoinJettonContract.address);
        logger = system.log(stablecoinJettonContract.address);
    });

    it("on repayBurned message should fail if amount repayed more than balance", async () => {
        await stablecoinJettonContract.send(
            stablecoinMaster,
            { value: toNano(1) },
            { $$type: "RepayBurnMessage", user: user.address, amount: 100n, fees: 1n }
        );
        console.log("stablecoinJettonContract", stablecoinJettonContract.address);

        await system.run();
        expect(track.collect()).toMatchSnapshot();
    });

    it("on repayBurned message should decrease balance and send repayBurnNotification to master", async () => {
        // increase balance before
        await stablecoinJettonContract.send(
            stablecoinMaster,
            { value: toNano(10) },
            {
                $$type: "TokenTransferInternal",
                queryId: 10n,
                amount: 1000n,
                from: user.address,
                responseAddress: null,
                forwardTonAmount: 0n,
                forwardPayload: beginCell().endCell(),
            }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();

        const walletData = await stablecoinJettonContract.getGetWalletData();
        const balance = walletData.balance;
        expect(balance).toEqual(1000n);

        await stablecoinJettonContract.send(
            stablecoinMaster,
            { value: toNano(1) },
            { $$type: "RepayBurnMessage", user: user.address, amount: 100n, fees: 1n }
        );

        await system.run();

        expect(track.collect()).toMatchSnapshot();

        const walletDataAfter = await stablecoinJettonContract.getGetWalletData();
        const balanceAfter = walletDataAfter.balance;
        expect(balanceAfter).toEqual(900n);
    });
});
