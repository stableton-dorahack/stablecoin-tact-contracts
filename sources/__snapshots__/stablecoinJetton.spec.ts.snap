// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StablecoinJettonContract on repayBurned message should decrease balance and send repayBurnNotification to master 1`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{178D4519000000000000000A203E8800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B204_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
          "to": "kQA5W0EgDzQi5RcifEJim7Hc2IKHXDcn2aFejFz4vefLKNCh",
          "type": "internal",
          "value": 10000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 4903n,
      },
    ],
  },
]
`;

exports[`StablecoinJettonContract on repayBurned message should decrease balance and send repayBurnNotification to master 2`] = `
[
  {
    "$seq": 2,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{B065C3AC800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B22C8203_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
          "to": "kQA5W0EgDzQi5RcifEJim7Hc2IKHXDcn2aFejFz4vefLKNCh",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 7561n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{8C5F4C71800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B22C8203_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQA5W0EgDzQi5RcifEJim7Hc2IKHXDcn2aFejFz4vefLKNCh",
            "to": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
            "type": "internal",
            "value": 991016000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`StablecoinJettonContract on repayBurned message should fail if amount repayed more than balance 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{B065C3AC800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B22C8203_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
          "to": "kQA5W0EgDzQi5RcifEJim7Hc2IKHXDcn2aFejFz4vefLKNCh",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "failed",
        "errorCode": 62972,
        "errorMessage": "Invalid balance",
      },
      {
        "$type": "sent-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQA5W0EgDzQi5RcifEJim7Hc2IKHXDcn2aFejFz4vefLKNCh",
          "to": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
          "type": "internal",
          "value": 995845000n,
        },
      },
    ],
  },
]
`;
